water/downloader.go

```diff
@@ -0,0 +1,118 @@
+package water
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+)
+
+//go:generate mockgen -destination=mock_downloader.go -package=water . WASMDownloader
+
+type WASMDownloader interface {
+	DownloadWASM(context.Context, io.Writer) error
+}
+
+type downloader struct {
+	urls            []string
+	httpClient      *http.Client
+	expectedHashSum string
+	httpDownloader  WASMDownloader
+}
+
+type DownloaderOption func(*downloader)
+
+func WithURLs(urls []string) DownloaderOption {
+	return func(d *downloader) {
+		d.urls = urls
+	}
+}
+
+func WithHTTPClient(httpClient *http.Client) DownloaderOption {
+	return func(d *downloader) {
+		d.httpClient = httpClient
+	}
+}
+
+func WithExpectedHashsum(hashsum string) DownloaderOption {
```

Comment:
Same as in the flashlight PR: we shouldn't need this independent reliability layer.